### PR TITLE
feat: [Context Switching] maintain focus of form control fields

### LINKED ISSUE #16684: [Task]: Maintain Context and Focus Phase 2
### Is there an existing issue for this?

- [X] I have searched the existing issues

## Entity Explorer: 
- [x] #17348
- [x] [Entity explorer should stay as you left it when you left the page (collapse levels)](https://github.com/appsmithorg/appsmith/issues/17975)
- [x] [Widget/explorer tab](https://github.com/appsmithorg/appsmith/issues/17975)
- [x] [Width - Should be the same across all pages of an app](https://github.com/appsmithorg/appsmith/issues/17975)
- [x] #15158

## [Canvas / Widgets](https://github.com/appsmithorg/appsmith/issues/17975):
- [x] Complex widgets, multi tier property panes
- [ ] Tabs
- [x] Property Pane width

## APIs:
- [x] #17435
- [x] #17436

## Queries: 
- [ ] ~Right side panel tabs~ No tabs on right unlike api pane

## Datasources
- [ ] ~#17437~
- [x] #17439
- [x] #17438

### LINKED ISSUE #17435: [Task]: Context switching - non code editors, we set focus but not cursor
Maintain cursor position for inputs.

- Pagination Fields

https://user-images.githubusercontent.com/12022471/199162083-04726785-d3b7-4e81-98c1-764ee47930cc.mov

- Settings inputs

https://user-images.githubusercontent.com/12022471/199162092-dc2197c4-f05c-435c-96e6-4470ed75cf9a.mov

- Switches

https://user-images.githubusercontent.com/12022471/199163886-162c51f6-0c72-484d-bf3d-8ea62d2e1889.mov

- Changing the query type in a dropdown leads to another property being selected when a user changes it, which can be confusing

https://user-images.githubusercontent.com/12022471/199164729-98622607-a48b-4148-833a-76d9c0fdeca1.mov

### PR DESCRIPTION
**Pull request by akash-codemonk**
## Description

Maintain focus of form control fields except in datasource forms. In addition we also add input focus to property pane fields.

Fixes #17435
Fixes #16684 

Code changes

- Replace soft focus with cursor focus assertion for code editor components https://github.com/appsmithorg/appsmith/pull/18240/files#diff-bede9c16af7e05a81e22bca222742814d8e5c86e2048fd9dc73fe7c0d779e30eL244
- Replaced `FocusElement.CodeEditor`, `FocusElement.PropertyField` with a single `FocusElement.ControlField` config to allow focusing on codeeditors as well as other types of fields. Name it `ControlField` as we use it to focus property controls and form controls used by query pane. https://github.com/appsmithorg/appsmith/pull/18240/files#diff-fe663a527556b1c6ff2ff44e8eb11b57815aad7dee6ebdcfac5a5d87c7ddadddR209
- Similar logic to focus form controls as used for property controls https://github.com/appsmithorg/appsmith/pull/18240/files#diff-06ac6baf58233321a0ca4fa3da4f64f17ddee57f7b8df5006c67f3b3b7506213R97

 

Media

https://user-images.githubusercontent.com/67054171/206709609-d8e5017d-ec98-4675-9206-2bd7ca04d11e.mov

## Type of change

- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

- Manual
- Cypress

### Issues raised during DP testing
https://github.com/appsmithorg/appsmith/issues/17435#issuecomment-1323419872

### Dev activity
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

### QA activity:
- [ ] Test plan has been approved by relevant developers
- [ ] Test plan has been peer reviewed by QA
- [ ] Cypress test cases have been added and approved by either SDET or manual QA
- [ ] Organized project review call with relevant stakeholders after Round 1/2 of QA
- [ ] Added Test Plan Approved label after reveiwing all Cypress test

### CODE REVIEW

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/DynamicHeight/DynamicHeight_Auto_Height_spec.js [hunk 1]
```diff
@@ -4,17 +4,15 @@ const commonlocators = require("../../../../locators/commonlocators.json");
 import { ObjectsRegistry } from "../../../../support/Objects/Registry";
 const agHelper = ObjectsRegistry.AggregateHelper;
 
-
-describe("Dynamic Height Width validation", function () {
-
+describe("Dynamic Height Width validation", function() {
```
Comment (hetunandu) [inline app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/DynamicHeight/DynamicHeight_Auto_Height_spec.js:7]: Add numbers on the test cases

"1. Dynamic Height.... "

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/IDE/MaintainContext&Focus_spec.js [hunk 2]
```diff
@@ -115,20 +116,18 @@ describe("MaintainContext&Focus", function() {
 
   it("5. Maintains focus on Query panes", () => {
     cy.SearchEntityandOpen("SQL_Query");
-    cy.assertCursorOnCodeInput(".t--actionConfiguration\\.body", {
-      ch: 5,
-      line: 0,
-    });
+    cy.get(".t--actionConfiguration\\.body .CodeEditorTarget").should(
```
Comment (hetunandu) [inline app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/IDE/MaintainContext&Focus_spec.js:119]: Why did we remove the cursor position assertion here?

File: app/client/src/navigation/FocusElements.ts [hunk 3]
```diff
@@ -151,17 +154,17 @@ export const FocusElementsConfig: Record<FocusEntity, Config[]> = {
     },
   ],
   [FocusEntity.QUERY]: [
-    {
```
Comment (hetunandu) [inline app/client/src/navigation/FocusElements.ts:154]: We still need to do focus and restore cursor on code editor fields

**Review by hetunandu — CHANGES_REQUESTED** [review]

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/DynamicHeight/DynamicHeight_Auto_Height_spec.js (hunk 1 above)
Comment (akash-codemonk) [reply app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/DynamicHeight/DynamicHeight_Auto_Height_spec.js:7]: This test wasn't added in this PR. The file was auto formatted by the pre-commit hook.

File: app/client/src/navigation/FocusElements.ts (hunk 3 above)
Comment (akash-codemonk) [reply app/client/src/navigation/FocusElements.ts:154]: This has been removed to remove input focus invoked by the CodeEditor component as it conflicts with the focus of the rest of the inputs with the form control fields. We do the same for codeeditor fields in the canvas for the property pane. 
Regardless the cursor is still maintained when the input gets focused(when the field is focused and we press enter we can see the restored cursor position) as this is stored directly by the codeeditor component itself identified by the datatree path. 


https://user-images.githubusercontent.com/67054171/205846740-c656bafb-7af5-40af-a1fe-ed032e62445a.mov

File: app/client/src/navigation/FocusElements.ts [hunk 4]
```diff
@@ -105,15 +101,14 @@ export enum FocusElement {
...
-  CodeEditor = "CodeEditor",
-  PropertyField = "PropertyField",
   PropertySections = "PropertySections",
   PropertyTabs = "PropertyTabs",
   PropertyPanelContext = "PropertyPanelContext",
   PropertyPaneWidth = "PropertyPaneWidth",
   SelectedPropertyPanel = "SelectedPropertyPanel",
   SelectedWidgets = "SelectedWidgets",
   SubEntityCollapsibleState = "SubEntityCollapsibleState",
+  ControlField = "ControlField",
```
Comment (hetunandu) [inline app/client/src/navigation/FocusElements.ts:111]: Can we rename it to "InputField" which is more generic and applies to all

**Review by hetunandu — CHANGES_REQUESTED** [review]

**Review by hetunandu — APPROVED** [review]

**Review by hetunandu — APPROVED** [review]

**Pull request merged**